### PR TITLE
Adds simultaneous support for in- and inter-process channels/servers

### PR DESCRIPTION
**Pull request by ST-DDT**
Adds #176 

This PR adds simultaneous support for in- and inter-process channels/servers.

----------------------------------

For the client side you just have to use the `inProcess:` prefix to target an inProcess service, at the same time you can also define the inter-process connections.

**Note:** This only works if you either use the default `GrpcChannelFactory` or wrap your own factory appropriately.

````properties
grpc.client.inProcess.address=inProcess:test
grpc.client.interProcess.address=static://localhost:9191
````

------------------------------------

For the server side you just have to use the new `grpc.server.inProcessName` property.

````properties
grpc.server.inProcessName=test
````

**Note:** The server auto configuration will only be used, if you don't define any of the server beans ( `GrpcServerFactory`/`GrpcServerLifecycle`) yourself.

------------------------------------

**Limitations**

Currently it is not possible to disable the inter-process server beans with property files alone. Using the server port 0 (random) will reduce the chance of collisions during tests that don't need the inter-process server. Another alternative could be creating a test configuration similar to `net.devh.boot.grpc.test.config.InProcessConfiguration` from the tests module. Maybe I'll improve that in the future.

### CODE REVIEW

File: grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * other) are only created on demand.
+ * </p>
+ */
+public class InProcessOrAlternativeChannelFactory implements GrpcChannelFactory {
+
+    private static final String IN_PROCESS_SCHEME = "inProcess";
+
+    private final GrpcChannelsProperties properties;
+    private final InProcessChannelFactory inProcessChannelFactory;
+    private final GrpcChannelFactory alternatveChannelFactory;
```
[the remark this answers is not in the document]
Comment (ST-DDT) [reply grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java]: Fixed. Travis seems to be broken though.

[the remark this answers is not in the document]
Comment (yidongnan) [reply grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java]: https://travis-ci.community/t/install-of-openjdk11-is-failing-again/3061/19
The master branch has been switched to xenial

File: grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * </ul>
+ *
+ * <p>
+ * Using this class does not incur any additional performance or resource costs, as the actual channels (in-process or
+ * other) are only created on demand.
+ * </p>
+ */
+public class InProcessOrAlternativeChannelFactory implements GrpcChannelFactory {
+
+    private static final String IN_PROCESS_SCHEME = "inProcess";
```
Comment (yidongnan) [inline grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java]: Is it better to use in-process as a scheme?

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerFactoryAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * The server lifecycle bean for the in-process-server.
+     *
+     * @param factory The factory used to create the lifecycle.
+     * @return The in-process server lifecycle bean.
+     */
+    @ConditionalOnBean(InProcessGrpcServerFactory.class)
+    @Bean
+    public GrpcServerLifecycle inProcessgrpcServerLifecycle(final InProcessGrpcServerFactory factory) {
```
Comment (yidongnan) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerFactoryAutoConfiguration.java]: `inProcessgrpcServerLifecycle` -> `inProcessGrpcServerLifecycle`

File: grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java (hunk 2 above)
Comment (ST-DDT) [reply grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java]: Oh, you a right. I didn't know that uri schemes are supposed to be lowercase. I'll fix it tomorrow.

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerAutoConfiguration.java [hunk 4]
```diff
@@ -103,34 +100,8 @@ public GrpcServerConfigurer decompressionServerConfigurer(final DecompressorRegi
...
-        for (final GrpcServiceDefinition service : serviceDiscoverer.findGrpcServices()) {
-            factory.addService(service);
-        }
-        return factory;
-    }
-
     @ConditionalOnMissingBean
+    @ConditionalOnBean(GrpcServerFactory.class)
     @Bean
     public GrpcServerLifecycle grpcServerLifecycle(final GrpcServerFactory factory) {
```
Comment (yidongnan) [inline grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerAutoConfiguration.java]: Is this still useful? Did I miss any?

Comment (ST-DDT) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerAutoConfiguration.java]: That one is used if the user specifies a custom `GrpcServerFactory` but does not specify a `GrpcServerLifecycle`. AFAICT some tests depent on/use this bean.

File: grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java (hunk 2 above)
Comment (ST-DDT) [reply grpc-client-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/client/channelfactory/InProcessOrAlternativeChannelFactory.java]: Fixed

File: grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerFactoryAutoConfiguration.java (hunk 3 above)
Comment (ST-DDT) [reply grpc-server-spring-boot-autoconfigure/src/main/java/net/devh/boot/grpc/server/autoconfigure/GrpcServerFactoryAutoConfiguration.java]: Fixed

**Pull request merged**